chore: update CONTRIBUTING to remove external PR (#573)
zama-ai/kms: CONTRIBUTING.md

There are two ways to contribute:

> - **Report issues**: Open issues on GitHub to report bugs, suggest improvements, or note typos.

Reviewer: ```suggestion
- **Report issues**: Open issues on GitHub to report bugs or suggest improvements.
```

> - **Propose code changes**: External contributors must work from their own fork of the repository. Do not push branches to `zama-ai/kms`, and do not open pull requests directly against `zama-ai/kms`.

Reviewer: Is this necessary to state? I mean, external contributors cannot push branches to our repos and forking is their only option yeah?

Reviewer 2: It's clearer to state it explicitly here I think :slightly_smiling_face:

Author: yes, seems obvious but we will have questions otherwise

Reviewer 3: still needs a CLA signed somewhere right ?

Reviewer 3: my bad it's later in the guide

> - **Propose code changes**: External contributors must work from their own fork of the repository. Do not push branches to `zama-ai/kms`, and do not open pull requests directly against `zama-ai/kms`.

Reviewer 4: I would start this way to make it more clear about what to do, and not only about what not to do

**Propose code changes**: To propose changes, first open an issue describing the problem and the proposed solution. Once the proposal has been reviewed and validated, fork the repository and develop your changes in a branch of your fork.

Note that external contributors are not allowed to push branches directly to zama-ai/kms or open pull request agains it.

In details, follow these steps:

* the problem you want to solve,
* the proposed approach,
* any expected impact on APIs, compatibility, security, deployment, or dependencies.

Zama will review the issue. If the proposed contribution is considered relevant, Zama may ask you to prepare the work in your fork. The contribution can then be integrated into `zama-ai/kms` after the required Contributor License Agreement (CLA) has been signed.

> Note that external contributors must work from their own fork of the repository. You could not open pull requests directly against `zama-ai/kms`.

Reviewer 5: ```suggestion
Note that external contributors must work from their own fork of the repository. You cannot directly open a pull request in `zama-ai/kms`.
```

## 1. Setting up the project

Start by [forking](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/fork-a-repo) the **KMS** repository.

External contributors must keep all development branches in their own fork. No branch should be pushed to the upstream `zama-ai/kms` repository.

> 

Reviewer: Same as above: what other workflow is possible for them? It's the only one, yes? I think we can remove this.

Reviewer 2: Why? Without this, the external contributor would have to guess the process right?

Reviewer: My two cents: If the only possible way to contribute is "the same process as any other git-based project you don't have commit rights to" then I think it's safe to omit. 
If a potential contributor has no idea of how to fork a repo and make a pull request upstream then there are better docs available elsewhere.

Author: the diff here is that there is no upstream PR as they dont have the rights to

- [ ] The PR title follows [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/), e.g. "chore: did something".
- [ ] Tests exist for all new `pub` methods and I checked that test coverage is at least as good as before this PR
- [ ] Code comments are in place for public methods and non-obvious code segments.

> - [ ] Any unfinished business is documented with a `TODO(#issue_number)` comment and brief description of what needs to be fixed.

Reviewer 2: Why this removal?

Author: up to the team but having an external contributor pushing code that open other issues seems weird, external PR should be atomic imho, I could restore that, let me know

Reviewer 2: Agreed. I think it makes sense to keep it somewhere if we go as Tore suggested though:

> Would it maybe make more sense to just separete this file into 2 sections? One for Zama employees, and one for external contributors?

Reviewer 5: We do already have this (in a condensed version) in https://github.com/zama-ai/kms/blob/main/.github/PULL_REQUEST_TEMPLATE.md

But indeed might be good to have a more detailed version somewhere, and maybe split things between internal and external contributions.

- [ ] The PR is _not_ updating _any_ dependencies (i.e. no changes to `Cargo.toml` or `Cargo.toml`), or if it does:
    - [ ] I confirm there are _only_ changes needed to fix compilation and tests (see [here](#Checklist-for-dependency-updates) for details.), or changes that happened due to bumping an internal version number.
- [ ] Changes in the PR do not affect the architecture of the protocol. Or if they do these steps must be taken:
    - [ ] A parallel PR or issue has been opened in the [tech-spec repo](https://github.com/zama-ai/tech-spec) (add the link here).
- [ ] Changes in the PR do not affect the architecture of the protocol or it should be clearly advertised and accepted in the Issue.
- [ ] The PR does not contain any breaking changes to the configuration and deployment files.
      (A change is _only_ considered breaking if a deployment configuration must be changed as part of an update. Adding new fields, with default values is _not_ considered breaking). Or if it does then these steps must be taken:
    - [ ] The PR is labeled with `devops`.

>     - [ ] The PR is marked using `!` in accordance with conventional commits. E.g. `chore!: changed decryption format according to Q3 release`.

Reviewer 2: I would have kept the `!` annotation.

Understood that other bullet points does not make sense for external contributor

Author: agreed will revert this one

    - [ ] An issue has been made for the relevant teams to handle these changes. 
- [ ] The PR does not contain changes to the interal protobufs/communication protocols that breaks compatibility with the previous version. That is, if we release from `main` after your PR, the KMS can still communicate with other KMS' that run the previous release. This ensures a much simpler upgrade process, in order to avoid the green/blue upgrade approach.
- [ ] The PR does not contain changes to the internal protobufs/communication protocols that breaks compatibility with the previous version. That is, if we release from `main` after your PR, the KMS can still communicate with other KMS' that run the previous release. This ensures a much simpler upgrade process, in order to avoid the green/blue upgrade approach.
- [ ] I have not changed existing `versionized` structs, nor added new `versionized` structs. Or if I have, these steps must be taken:
    - [ ] The backwards compatibility tests have been updated and/or new tests covering the changes have been added.
- [ ] The PR does not contain changes to the critical business logic or cryptographic code. Or if it does then these steps must be taken:

>     - [ ] At least two people must be assigned as reviewers and eventually approve the PR.

Reviewer 2: Why this removal?

Author: this looks more like internal PR guidance / PR template than contributing guide imho

We currently consider this an acceptable risk. Hence there is no need to manually modify `Cargo.lock`.

> ## 7. Proposing a contribution

Reviewer: I prefer the old wording.

Reviewer 2: Old wording is a bit misleading as we forbid pull request on our repo...

Reviewer: We do? They can still propose a PR like a normal OSS project, i.e. from their own fork right?

Author: no they wont be able, we will disable this feature

> Once your changes are ready, push the branch to your fork.

Reviewer: Here too.

Reviewer 2: I think it's clearer now :sweat_smile:

Do not open a pull request directly against `zama-ai/kms`.

Instead, update the related issue with:
* a link to your fork, branch, and optionally PR
* a short summary of the changes,
* test results,
* any relevant notes for reviewers.

Opening a PR in your fork (from your local `branch` into your local `main`) is an option if you want to discuss implementation. In this case, add the PR link in the related issue.

> If Zama decides to integrate the contribution, Zama will coordinate the next steps. Before integration, you must sign the Contributor License Agreement (CLA). Our CLA-bot will guide you through this process when a pull request is opened on `zama-ai/kms`.

Reviewer 4: If the contribution is selected for integration, contributors will be asked to sign the Contributor License Agreement (CLA). The CLA bot will guide you through the process when a pull request is opened on `zama-ai/kms`.

Reviewer 3: just a tiny bit unclear on the "opening a pr on zama-ai/kms" given external contributor (with my current understanding) cannot do that

does it mean we need the thing shared by Thomas on vouch/denounce from ghostty to make this work ?

Author: the PR will be open by a Zama dev but CLA bot will find the main author and complains if it's not in the list
